### PR TITLE
Add cli config flag

### PR DESCRIPTION
**Pull request by gouglhupf**
*Issue #11 

*Description of changes:* Adds support for  parameters/tags as a json file for the deploy command. See the issue for more details on why this may be useful.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by gouglhupf** [comment]
@stilvoid Ping :). Would it be possible to have a look at these changes and post your thoughts?

**Comment by stilvoid** [comment]
Hey, thanks so much for this and apologies for the delay in getting back to you!

I like the approach and this definitely fits what I had in mind but I have a few minor changes I'd suggest which I'll put into a review on this PR in a moment.

File: internal/cmd/deploy/deploy.go [hunk 1]
```diff
@@ -33,7 +42,23 @@ If you don't specify a stack name, rain will use the template filename minus its
 
 If a template needs to be packaged before it can be deployed, rain will package the template first.
 Rain will attempt to create an S3 bucket to store artifacts that it packages and deploys.
-The bucket's name will be of the format rain-artifacts-<AWS account id>-<AWS region>`,
+The bucket's name will be of the format rain-artifacts-<AWS account id>-<AWS region>.
+
+The CLI JSON file (--cli-json-file <string>) can be used to programmatically set tags and parameters.
```
Comment (stilvoid) [inline internal/cmd/deploy/deploy.go:47]: `--cli-json-file` seems a bit long-winded to me. I'd prefer we take even more inspiration from CodePipline and refer to this file as "template configuration" and use `-c` and `--config` for it.

File: internal/cmd/deploy/deploy.go [hunk 2]
```diff
@@ -61,6 +86,33 @@ The bucket's name will be of the format rain-artifacts-<AWS account id>-<AWS reg
 		// Parse params
 		parsedParams := ListToMap("param", params)
 
+		if len(cliJSONFile) != 0 {
+			jsonFile, err := os.Open(cliJSONFile)
```
Comment (stilvoid) [inline internal/cmd/deploy/deploy.go:90]: Could just use `ioutil.ReadFile` here :)

File: internal/cmd/deploy/deploy.go [hunk 3]
```diff
@@ -61,6 +86,33 @@ The bucket's name will be of the format rain-artifacts-<AWS account id>-<AWS reg
...
+			}
+			defer jsonFile.Close()
+
+			byteValue, err := ioutil.ReadAll(jsonFile)
+			if err != nil {
+				panic(ui.Errorf(err, "unable to read cli json file '%s'", cliJSONFile))
+			}
+
+			var cliInput cliInputJSON
+			err = json.Unmarshal([]byte(byteValue), &cliInput)
```
Comment (stilvoid) [inline internal/cmd/deploy/deploy.go:102]: One of rain's goals is to not care what format CloudFormation templates you prefer, I think we should do the same with the configuration file and use the yaml parser (JSON is valid YAML).

File: internal/cmd/deploy/deploy.go [hunk 4]
```diff
@@ -61,6 +86,33 @@ The bucket's name will be of the format rain-artifacts-<AWS account id>-<AWS reg
...
+			if err != nil {
+				panic(ui.Errorf(err, "unable to parse json in '%s'", cliJSONFile))
+			}
+
+			for k, v := range cliInput.Tags {
+				parsedTags[k] = v
+			}
+
+			for k, v := range cliInput.Parameters {
+				parsedParams[k] = v
```
Comment (stilvoid) [inline internal/cmd/deploy/deploy.go:112]: This means that tags and parameters specified in a config file will override tags and parameters supplied by `--tags` or `--params` which seems counter-intuitive.

Perhaps we should warn (or fail) on clashing tags/param keys?

**Review by stilvoid — CHANGES_REQUESTED** [review]
Thanks again for the contribution!

**Comment by gouglhupf** [comment]
Thanks for the feedback. I just pushed new changes to address your comments.

File: internal/cmd/deploy/deploy.go (hunk 1 above)
Comment (gouglhupf) [reply internal/cmd/deploy/deploy.go]: True, `-c` and `--config` is more concise!

File: internal/cmd/deploy/deploy.go (hunk 3 above)
Comment (gouglhupf) [reply internal/cmd/deploy/deploy.go]: Good idea, I changed it to the YAML parser.

File: internal/cmd/deploy/deploy.go (hunk 4 above)
Comment (gouglhupf) [reply internal/cmd/deploy/deploy.go]: I changed the priority. The cli flags `tag` and `param` can now override the config file. When they do so, a warning will be printed. Is this what you had in mind?

**Pull request merged**